src/osd/PrimaryLogPG.cc

```diff
@@ -3457,24 +3458,121 @@ void PrimaryLogPG::cancel_manifest_ops(bool requeue, vector<ceph_tid_t> *tids)
   }
 }
 
-void PrimaryLogPG::refcount_manifest(ObjectContextRef obc, object_locator_t oloc, hobject_t soid,
-                                     SnapContext snapc, bool get, RefCountCallback *cb, uint64_t offset)
+void PrimaryLogPG::dec_refcount(ObjectContextRef obc, const object_info_t& oi, 
+				const object_ref_delta_t& refs)
+{
+  for (auto p = refs.begin(); p != refs.end(); ++p) {
+    dout(10) << __func__ << ": decrement reference on offset oid: " << p->first << dendl;
+    object_locator_t target_oloc(p->first);
+    refcount_manifest(obc, obc->obs.oi.soid, target_oloc, p->first, 
+		      SnapContext(), refcount_t::DECREMENT_REF, NULL);
+  }
+}
+
+
+void PrimaryLogPG::dec_all_refcount_head_manifest(object_info_t& oi, OpContext* ctx, const hobject_t &coid)
+{
+  SnapSetContext* ssc = ctx->obc->ssc;
+  ceph_assert(oi.has_manifest());
+  ceph_assert(oi.soid.is_head());
+  // has snapshot
+  if (ssc && ssc->snapset.clones.size() > 0) {
+    ceph_assert(ssc);
+    dout(15) << __func__ <<  " has snapset " << dendl;
+    object_ref_delta_t refs;
+    ceph_assert(oi.manifest.is_chunked());
+    ceph_assert(!oi.manifest.is_redirect());
+
+    SnapSet& snapset = ssc->snapset;
+
+    // check adjacent clones
+    auto s = std::find(snapset.clones.begin(), snapset.clones.end(), coid.snap);
```

Comment:
You are passing coid in as hobject_t() for head.  That means that snap is actually 0 rather than NO_SNAP.  I don't understand why you don't just pass in the actual hobject_t in that case.  It happens to sort of work because no clone will have snapid 0.

Reply:
Or simply use oi.soid -- oi should always be the object_info_t for the object being removed.